Tom/filip (#27)
streambased-io/streambased-demos: cases/6_all_in_one/README.md

# Streambased All In One Demo

This demo shows the full complement of Streambased technologies using simple, inch deep examples. For more business 
case focused examples please see the other demos in this repository.

This demo will showcase:

1. Accessing Kafka data as if it was an S3 filesystem
2. Accessing Kafka data as if it was an Iceberg Table
3. Accessing Kafka data as if it was ANSI SQL compatible database.

All of the above access patterns will utilise Streambased indexing to accelerate performance and all will only use 
Kafka as a data source (no ETL is involved in any of these demos)

## Environment

> This demo simulates a typical operational environment in Kafka, we are simulating a shopify style 
> retail company where users host shops for their goods on our platform. The architecture looks 
> like this:
> 
> All of this is exists in a single docker-compose environment containing the following core Kafka services:

Reviewer: Doesnt sound right to me... imho.

```suggestion
This demo simulates an operational environment in Kafka, we are simulating a shopify style 
retail company where users host shops for their goods on our platform. The demo is deployed as a single docker-compose environment comprising of:

Core Kafka services:
```

view topics as if they were S3 buckets and will see a number of files sliced by offset chunk inside. Streambased uses 
a naming convention to easily identify the contents of the file. For instance a file that represents offsets 100-1100 
of partition 0 would be named `0-100-1100.avro`. Streambased also uses a directory structure to represent the chunking 
used so that 1 topic can be chunked in multiple ways. For instance `/1000/0-100-1100.avro` is a file in a directory 
representing chunks of 1000 offsets. Similarly `/100/0-100-200.avro` represents 100 offset chunking. 

Note that these files do not actually exist and hold no data. Only when they are downloaded by a client is the data 
fetched from Kafka and returned.

We can use the packaged s3manager client to explore our topics. Navigate to `localhost:9180` to see our S3 buckets:

![buckets](media/buckets.png "Buckets")

Within each bucket we see files corresponding to Kafka data, these can be downloaded and worked with locally. 

![files](media/files.png "Files")

### As an Iceberg table.

The main reason for the above functionality is to support Apache Iceberg tables. Iceberg consists of a set of data 

> files and a set of metadata files tht describe access to them. We have already created the data files above and 

Reviewer: ```suggestion
files and a set of metadata files that describe access to them. We have already created the data files above and 
```
